### PR TITLE
ENH: more numerically stable version of `softmax`

### PR DESCRIPTION
**Pull request by pmli**
#### Reference issue
There is no issue that this PR closes.

#### What does this implement/fix?
This implements a more accurate version of `softmax` and makes some improvements to the docstrings in `_logsumexp.py`

#### Additional information
This is based on the following paper: Pierre Blanchard, Desmond J. Higham, Nicholas J. Higham, **Accurately computing the log-sum-exp and softmax functions**, *IMA Journal of Numerical Analysis*, Volume 41, Issue 4, October 2021, Pages 2311–2330, [DOI](https://doi.org/10.1093/imanum/draa038). There is also an [online seminar](https://media.ed.ac.uk/playlist/dedicated/51612401/1_ttaqxfs9/1_inp881ig) (the first half is about the paper) given by Des Higham and a [blog post](https://nhigham.com/2020/02/03/accurately-computing-the-softmax-function) by Nick Higham.

The paper derives error bounds for different implementations of `logsumexp` and `softmax`, indicating which implementations are more accurate. Then it demonstrates the theory on an example from deep learning. To make it easier, I reproduced the results with a randomly-generated benchmark. Here is the code demonstrating different implementations of `logsumexp`, `softmax`, and `log_softmax`: [logsumexp_test.py.txt](https://github.com/scipy/scipy/files/8693862/logsumexp_test.py.txt) (".txt" should be removed). The result is the following figure:

comparison

The following are specific remarks:

- `logsumexp`:
  - SciPy uses the "shifted" implementation.
  - The paper discusses the naive and "log1p" implementations ("log1p" is like "shifted", with the additional trick of using `log1p` after shifting).
  - The paper argues that the "log1p" implementation avoids overflow and is of similar accuracy to the naive implementation when there is no overflow.
  - The figure shows that the different implementations of `logsumexp` are of similar accuracy to the naive implementation (when there is no overflow).
  - I didn't implement the "log1p" version because: 1. I don't know how to do it for multi-dimensional arrays, 2. it doesn't seem to be worth the effort in terms of accuracy, and 3. the implementation of `logsumexp` is already somewhat complex because of the support for weights (where the `log1p` trick cannot be used in general).
- `softmax`:
  - SciPy uses the "alternative" implementation.
  - The paper discusses all three implementations.
  - The paper argues that the "shifted" implementation avoids overflow, it is of similar accuracy to the naive implementation when there is no overflow, and it is more accurate than the "alternative" implementation.
  - The figure confirms the results of the paper.
  - Changing SciPy's implementation also improved the example in the docstring.
- `log_softmax`:
  - SciPy uses the "alternative2" implementation (it is similar to "alternative", except that it avoids an additional shifting).
  - The paper doesn't discuss `log_softmax`.
  - The figure shows that SciPy's implementation is accurate enough, although the implementation could be simplified at the cost of an additional shifting.

There was a discussion happening in a PR [here](https://github.com/scipy/scipy/pull/8872#discussion_r192577457), but before the above paper was published. Interestingly, [`scikit-learn` had implemented](https://github.com/scikit-learn/scikit-learn/pull/5225) the more accurate version of `softmax`.

### CODE REVIEW

**Review by person142 — CHANGES_REQUESTED** [review]
Changes look good, but @pmli can you add the paper as a reference in the docstring?

**Comment by pmli** [comment]
Thank you @person142. I've added the reference now. Does it look ok?

**Review by person142 — APPROVED** [review]
LG, tests still running, ping me if I forget to merge.

**Comment by pmli** [comment]
Hi @person142, the tests seem to have passed successfully.

**Comment by person142** [comment]
Thanks for the reminder!

**Comment by pmli** [comment]
Thanks for the review!
